inlong-sort/sort-connectors/pulsar/src/main/java/org/apache/inlong/sort/pulsar/table/DynamicPulsarDeserializationSchema.java

```diff
@@ -0,0 +1,347 @@
+
+    private final TypeInformation<RowData> producedTypeInfo;
+
+    private final boolean upsertMode;
+
+    private SourceMetricData sourceMetricData;
+
+    private String inlongMetric;
+
+    private static final ThreadLocal<SimpleCollector<RowData>> tlsCollector =
+            new ThreadLocal<SimpleCollector<RowData>>() {
+                @Override
+                public SimpleCollector initialValue() {
+                    return new SimpleCollector();
+                }
+            };
+
+    DynamicPulsarDeserializationSchema(
+            int physicalArity,
+            @Nullable DeserializationSchema<RowData> keyDeserialization,
+            int[] keyProjection,
+            DeserializationSchema<RowData> valueDeserialization,
+            int[] valueProjection,
+            boolean hasMetadata,
+            MetadataConverter[] metadataConverters,
+            TypeInformation<RowData> producedTypeInfo,
+            boolean upsertMode,
+    String inlongMetric) {
+        if (upsertMode) {
+            Preconditions.checkArgument(
+                    keyDeserialization != null && keyProjection.length > 0,
+                    "Key must be set in upsert mode for deserialization schema.");
+        }
+        this.keyDeserialization = ThreadSafeDeserializationSchema.of(keyDeserialization);
+        this.valueDeserialization = ThreadSafeDeserializationSchema.of(valueDeserialization);
+        this.hasMetadata = hasMetadata;
+        this.outputCollector = new OutputProjectionCollector(
+                physicalArity,
+                keyProjection,
+                valueProjection,
+                metadataConverters,
+                upsertMode);
+        this.producedTypeInfo = producedTypeInfo;
+        this.upsertMode = upsertMode;
+        this.inlongMetric = inlongMetric;
+    }
+
+    @Override
+    public void open(DeserializationSchema.InitializationContext context) throws Exception {
+        if (keyDeserialization != null) {
+            keyDeserialization.open(context);
+        }
+        valueDeserialization.open(context);
+
+        sourceMetricData = new SourceMetricData(context.getMetricGroup());
+        if (inlongMetric != null && !inlongMetric.isEmpty()) {
+            String[] inLongMetricArray = inlongMetric.split(DELIMITER);
+            String groupId = inLongMetricArray[0];
+            String streamId = inLongMetricArray[1];
+            String nodeId = inLongMetricArray[2];
+            sourceMetricData.registerMetricsForNumBytesIn(groupId,
+                streamId, nodeId, NUM_BYTES_IN);
+            sourceMetricData.registerMetricsForNumBytesInPerSecond(groupId,
+                streamId, nodeId, NUM_BYTES_IN_PER_SECOND);
+            sourceMetricData.registerMetricsForNumRecordsIn(groupId, streamId,
+                nodeId, NUM_RECORDS_IN);
+            sourceMetricData.registerMetricsForNumBytesInPerSecond(groupId, streamId,
+                nodeId, NUM_RECORDS_IN_PER_SECOND);
```

Comment:
modify `registerMetricsForNumBytesInPerSecond` to `registerMetricsForNumRecordsInPerSecond`